From 47a394a3c4ed5d623ba51ab7258cbc31569bc039 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Wed, 10 Apr 2024 11:58:43 +0200
Subject: [PATCH 0756/1695] drivers: regulator: fix error case in
 regulator_supported_voltages()

Fix case when .supported_voltages() handler is supported to assert
voltage description when the handler succeeds and to use pre-filled
fallback description when the handler returns TEE_ERROR_NOT_SUPPORTED.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git cfc4439aefb731244db6b1c623dbebcb7ccdc02c drivers: regulator: fix error case in regulator_supported_voltages()]
Change-Id: I20e38550d5a9c0fd79fbf0bcd8377b656b009528
Fixes: af5b9881111c ("drivers: regulator: supported voltage consider levels bounds")
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/372699
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/regulator/regulator.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/core/drivers/regulator/regulator.c b/core/drivers/regulator/regulator.c
index 92f63f3a0..477b8c158 100644
--- a/core/drivers/regulator/regulator.c
+++ b/core/drivers/regulator/regulator.c
@@ -249,18 +249,18 @@ TEE_Result regulator_supported_voltages(struct regulator *regulator,
 					struct regulator_voltages_desc **desc,
 					const int **levels)
 {
-	assert(regulator && desc && levels);
+	TEE_Result res = TEE_ERROR_NOT_SUPPORTED;
 
-	if (regulator->ops->supported_voltages) {
-		TEE_Result res = TEE_ERROR_GENERIC;
+	assert(regulator && desc && levels);
 
+	if (regulator->ops->supported_voltages)
 		res = regulator->ops->supported_voltages(regulator, desc,
 							 levels);
-		if (res != TEE_ERROR_NOT_SUPPORTED)
-			return res;
-	} else {
+	if (res == TEE_ERROR_NOT_SUPPORTED) {
 		*desc = &regulator->voltages_fallback.desc;
 		*levels = regulator->voltages_fallback.levels;
+	} else if (res) {
+		return res;
 	}
 
 	assert(((*desc)->type == VOLTAGE_TYPE_FULL_LIST &&
-- 
2.47.2

